File: TripKitAndroid/src/androidTest/java/com/skedgo/tripkit/ServiceResponseTest.java
package com.skedgo.tripkit;

import android.content.Context;
import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.runner.AndroidJUnit4;
import com.google.gson.Gson;
import com.google.gson.stream.JsonReader;
import com.skedgo.tripkit.common.model.Location;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

import org.junit.Before;
import org.junit.runner.RunWith;
import com.skedgo.tripkit.routing.RealTimeVehicle;
import com.skedgo.tripkit.routing.Shape;

import static org.assertj.core.api.Java6Assertions.assertThat;

@RunWith(AndroidJUnit4.class)
public class ServiceResponseTest  {
  private Context context;

  @Before
  public void setUp() {
    // InstrumentationRegistry will do the trick!
    // We can use this context to test implementations of SQLiteOpenHelper.
    context = InstrumentationRegistry.getInstrumentation().getTargetContext();
  }


  public void testSerialize() throws IOException {
    final InputStream stream = context.getAssets().open("ServiceResponseTest.json");
    final ServiceResponse response = new Gson().fromJson(
        new JsonReader(new InputStreamReader(stream)),
        ServiceResponse.class
    );

    assertThat(response).isNotNull();
    assertThat(response.realTimeStatus()).isEqualTo("CAPABLE");
    assertThat(response.shapes())
        .isNotNull()
        .hasSize(1)
        .doesNotContainNull();
    final Shape firstShape = response.shapes().get(0);
    assertThat(firstShape.getStops())
        .isNotNull()
        .hasSize(26);
  }

  public void testUpdatedExample() throws IOException {
    final InputStream stream = context.getAssets().open("ServiceResponseTestStops.json");
    final ServiceResponse response = new Gson().fromJson(
        new JsonReader(new InputStreamReader(stream)),
        ServiceResponse.class
    );

    assertThat(response).isNotNull();
    assertThat(response.realTimeStatus()).isEqualTo("CAPABLE");
    assertThat(response.shapes())
        .isNotNull()
        .hasSize(1)
        .doesNotContainNull();
    final Shape firstShape = response.shapes().get(0);
    assertThat(firstShape.getStops())
        .isNotNull()
        .hasSize(58);
  }

  public void testRealTimeVehicle() throws IOException {
    final InputStream stream = context.getAssets().open("ServiceResponseTestRealTime.json");
    final ServiceResponse response = new Gson().fromJson(
        new JsonReader(new InputStreamReader(stream)),
        ServiceResponse.class
    );

    assertThat(response).isNotNull();
    assertThat(response.realTimeStatus()).isEqualTo("IS_REAL_TIME");
    assertThat(response.shapes())
        .isNotNull()
        .hasSize(1)
        .doesNotContainNull();
    final Shape firstShape = response.shapes().get(0);
    assertThat(firstShape.getStops())
        .isNotNull()
        .hasSize(41);

    RealTimeVehicle vehicle = response.realtimeVehicle();

    assertThat(vehicle).isNotNull();
    assertThat(vehicle.getLastUpdateTime()).isEqualTo(1468594846);
    assertThat(vehicle.getLocation()).isNotNull();

    Location location = vehicle.getLocation();

    assertThat(location.getLat()).isEqualTo(-33.86224);
    assertThat(location.getLon()).isEqualTo(151.21207);
    assertThat(location.getBearing()).isEqualTo(6);

    assertThat(response.realtimeAlternativeVehicle())
        .isNotNull()
        .hasSize(1)
        .doesNotContainNull();

  }
}

File: TripKitAndroid/src/test/java/com/skedgo/tripkit/serviceresponse/ServiceResponseTest.kt
package com.skedgo.tripkit.serviceresponse

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.google.gson.Gson
import com.skedgo.tripkit.ServiceResponse
import com.skedgo.tripkit.booking.ui.base.MockKTest
import com.skedgo.tripkit.mock.MockServiceResponse
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import java.io.IOException

@RunWith(AndroidJUnit4::class)
class ServiceResponseTest: MockKTest() {

    @get:Rule
    val rule = InstantTaskExecutorRule()

    @Test
    @Throws(IOException::class)
    fun testSerialize() {
        val response = Gson().fromJson(
            MockServiceResponse.SERVICE_RESPONSE_TEST,
            ServiceResponse::class.java
        )

        assertThat(response).isNotNull
        assertThat(response.realTimeStatus()).isEqualTo("CAPABLE")
        assertThat(response.shapes())
            .isNotNull
            .hasSize(1)
            .doesNotContainNull()

        val firstShape = response.shapes()[0]
        assertThat(firstShape.stops)
            .isNotNull
            .hasSize(26)
    }

    @Test
    @Throws(IOException::class)
    fun testUpdatedExample() {
        val response = Gson().fromJson(
            MockServiceResponse.SERVICE_RESPONSE_TEST_STOPS,
            ServiceResponse::class.java
        )

        assertThat(response).isNotNull
        assertThat(response.realTimeStatus()).isEqualTo("CAPABLE")
        assertThat(response.shapes())
            .isNotNull
            .hasSize(1)
            .doesNotContainNull()

        val firstShape = response.shapes()[0]
        assertThat(firstShape.stops)
            .isNotNull
            .hasSize(58)
    }

    @Test
    @Throws(IOException::class)
    fun testRealTimeVehicle() {
        val response = Gson().fromJson(
            MockServiceResponse.SERVICE_RESPONSE_TEST_REAL_TIME,
            ServiceResponse::class.java
        )

        assertThat(response).isNotNull
        assertThat(response.realTimeStatus()).isEqualTo("IS_REAL_TIME")
        assertThat(response.shapes())
            .isNotNull
            .hasSize(1)
            .doesNotContainNull()

        val firstShape = response.shapes()[0]
        assertThat(firstShape.stops)
            .isNotNull
            .hasSize(41)

        val vehicle = response.realtimeVehicle()

        assertThat(vehicle).isNotNull
        assertThat(vehicle?.lastUpdateTime).isEqualTo(1468594846)
        assertThat(vehicle?.location).isNotNull

        val location = vehicle?.location

        assertThat(location?.lat).isEqualTo(-33.86224)
        assertThat(location?.lon).isEqualTo(151.21207)
        assertThat(location?.bearing).isEqualTo(6)

        assertThat(response.realtimeAlternativeVehicle())
            .isNotNull
            .hasSize(1)
            .doesNotContainNull()
    }
}
